### PR TITLE
implement ProgressWriter

### PR DESCRIPTION
**Pull request by mxyng**
Simplify writing progress by implementing a ProgressWriter. Using `io.TeeReader`, the ProgressWriter will be responsible for updating user on the progress of the upload (and eventually down) without manually defining copy chunks.

### CODE REVIEW

File: server/upload.go [hunk 1]
```diff
@@ -121,8 +91,6 @@ func uploadBlobChunked(ctx context.Context, requestURL *url.URL, layer *Layer, r
 
 			switch {
 			case resp.StatusCode == http.StatusUnauthorized:
-				ch <- errors.New("unauthorized")
```
Comment (BruceMacD) [inline server/upload.go:124]: what happens to the progress writer on error?

File: server/upload.go [hunk 2]
```diff
@@ -170,3 +138,23 @@ func uploadBlobChunked(ctx context.Context, requestURL *url.URL, layer *Layer, r
...
+func (pw *ProgressWriter) Write(p []byte) (n int, err error) {
+	pw.completed += len(p)
+	pw.fn(api.ProgressResponse{
+		Status:    pw.status,
+		Digest:    pw.digest,
+		Total:     pw.total,
+		Completed: pw.completed,
+	})
+
+	return 0, nil
```
Comment (BruceMacD) [inline server/upload.go:159]: does returning 0 here indicate nothing was written?

Comment (mxyng) [reply server/upload.go:159]: Good catch. It should be len(p) though it doesn't seem to make a difference.

File: server/upload.go (hunk 1 above)
Comment (mxyng) [reply server/upload.go:124]: This was used to cleanup the goroutine. Since that goroutine doesn't exist, it no longer needed

Comment (mxyng) [reply server/upload.go:124]: Added a separate error handler for max retries, which seemed to be missing from main

**Review by jmorganca — APPROVED** [review]

File: server/upload.go [hunk 3]
```diff
@@ -170,3 +111,72 @@ func uploadBlobChunked(ctx context.Context, requestURL *url.URL, layer *Layer, r
...
+	fn        func(api.ProgressResponse)
+}
+
+func (pw *ProgressWriter) Write(b []byte) (int, error) {
+	n := len(b)
+	pw.bucket += n
+	pw.completed += n
+
+	// throttle status updates to not spam the client
+	if pw.bucket >= 1024*1024 || pw.completed >= pw.total {
```
Comment (mxyng) [inline server/upload.go:170]: Write happens in sizes of 32KB so this will batch up 32 write requests into a single progress update

**Pull request merged**